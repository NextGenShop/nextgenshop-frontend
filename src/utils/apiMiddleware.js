import axios from "axios";
import config from "../../config";

// exampleRequest {
//   headers?: any;
//   data?: any;
//   url: string;
//   method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
// }

const apiMiddleware = ({ dispatch }) => (next) => (action) => {
  next(action);

  if (action.type !== "API") return;

  const { name, url, requestData, extraData } = action;
  axios.defaults.baseURL = config.API_URL;
  axios.defaults.headers.common["Content-Type"] = "application/json";

  let req = {
    url,
    method: requestData.method,
  };
  if (requestData.headers) {
    req.headers = requestData.headers;
  }
  if (requestData.data) {
    req.data = requestData.data;
  }

  dispatch({ type: `${name}_REQUEST` });
  axios
    .request(req)
    .then(({ data }) => {
      dispatch({ type: `${name}_SUCCESS`, response: data, extraData });
    })
    .catch((err) => {
      const errorMessage = err.response ? err.response.data.error : "An unknown error has occurred";
      dispatch({ type: `${name}_FAILURE`, error: errorMessage });
    });
};

export default apiMiddleware;
