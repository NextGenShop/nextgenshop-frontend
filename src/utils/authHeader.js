export default function authHeader(state) {
  const token = state.auth.token;
  if (token) {
    return { authorization: `Token ${token}` };
  } else {
    return null;
  }
}
