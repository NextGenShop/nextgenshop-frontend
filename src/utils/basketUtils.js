const calcTotalPrice = (items) => {
  let totalPrice = 0;
  for (let item of items) {
    totalPrice += item.product.price * item.quantity;
  }
  return totalPrice;
};

const addBasketItem = (product, basket, quantity = 1) => {
  let newItems = [];
  const foundItem = basket.items.find(
    (item) => item.product.productId === product.productId
  );
  if (foundItem) {
    newItems = basket.items.map((item) =>
      item.product.productId === foundItem.product.productId
        ? { ...item, quantity: item.quantity + quantity }
        : item
    );
  } else {
    newItems = [...basket.items, { product, quantity }];
  }
  return { items: newItems };
};

const removeBasketItem = (productId, basket) => {
  let newItems = basket.items.map((item) => {
    if (item.product.productId === productId) {
      if (item.quantity <= 1) {
        return null;
      } else {
        return { ...item, quantity: item.quantity - 1 };
      }
    }
    return item;
  });
  newItems = newItems.filter((item) => item !== null);
  return { items: newItems };
};

export { addBasketItem, removeBasketItem, calcTotalPrice };
