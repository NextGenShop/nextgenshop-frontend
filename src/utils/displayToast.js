import { toastr } from "react-redux-toastr";

let options = {
  timeOut: 5000,
  showCloseButton: true,
  progressBar: false,
  transitionIn: "fadeIn",
  transitionOut: "fadeOut",
};

export const displayToast = (title, msg, type, callbackClear) => {
  const toastrInstance =
    type === "info"
      ? toastr.info
      : type === "warning"
      ? toastr.warning
      : type === "danger"
      ? toastr.error
      : toastr.success;
  if (callbackClear) options.onShowComplete = callbackClear;
  toastrInstance(title, msg, options);
};
