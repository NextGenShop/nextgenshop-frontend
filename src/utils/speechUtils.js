import recognizeMic from "watson-speech/speech-to-text/recognize-microphone";
import synthesize from "watson-speech/text-to-speech/synthesize";

//audio/wav is supported in Chrome, Edge, Firefox (v3.5), Opera (v11.00), Safari (v3.1)
//audio/mp3 is supported in Chrome, IE9, Edge, Firefox (v71), Opera, Safari (v3.1)
//ref: https://en.wikipedia.org/wiki/HTML5_audio

export const getSupportedAudioFormat = (audioElement) => {
  if (!audioElement) {
    audioElement = document.createElement("audio");
  }

  let accept =
    audioElement &&
    audioElement.canPlayType === "function" &&
    audioElement.canPlayType("audio/mp3") !== ""
      ? "audio/mp3"
      : "audio/wav";

  return accept;
};

export const captureAudioFromMicrophone = async (
  speechToTextUrl,
  speechToTextToken
) => {
  let mediaStream = null;
  try {
    mediaStream = await navigator.mediaDevices.getUserMedia({
      video: false,
      audio: true,
    });
  } catch (err) {
    console.log(err);
  }

  const recognizeMicrophoneStream = recognizeMic({
    url: speechToTextUrl,
    accessToken: speechToTextToken,
    objectMode: true,
    extractResults: true,
    format: true,
    mediaStream,
    keepMicrophone: true,
  });

  return recognizeMicrophoneStream;
};

export const textToSpeech = async (
  textToSpeechUrl,
  textToSpeechToken,
  text,
  voice,
  accept,
  element
) => {
  return synthesize({
    url: textToSpeechUrl,
    accessToken: textToSpeechToken,
    text,
    voice,
    accept,
    element,
  });
};

export const createAssistantSession = async (assistantUrl, assistantToken) => {
  let session;
  try {
    const res = await fetch(`${assistantUrl}/sessions?version=2020-04-01`, {
      method: "post",
      headers: new Headers({
        Authorization: "Bearer " + assistantToken,
        "Content-Type": "application/json",
      }),
    });
    session = await res.json();
  } catch (err) {
    console.log(err);
  }
  return session ? session.session_id : null;
};

export const messageAssistant = async (
  assistantUrl,
  assistantToken,
  sessionId,
  message
) => {
  let res;
  let data = { speech: "", action: null };
  try {
    res = await fetch(
      `${assistantUrl}/sessions/${sessionId}/message?version=2020-04-01`,
      {
        method: "post",
        headers: new Headers({
          Authorization: "Bearer " + assistantToken,
          Accept: "application/json",
          "Content-Type": "application/json",
        }),
        body: JSON.stringify({ input: { text: message } }),
      }
    );
    res = await res.json();
    console.log(res);
    data.speech =
      '<speak version="1.0"><prosody rate="+15%">' +
      res.output.generic
        .reduce((acc, obj) => {
          if (obj.response_type === "text") {
            acc.push(obj.text);
          }
          return acc;
        }, [])
        .join(' <break strength="medium"></break>') +
      "</prosody></speak>";
    const actions = res.output.generic.reduce((acc, obj) => {
      if (obj.response_type === "user_defined") {
        if (obj.user_defined.quantity) {
          obj.user_defined.action_type = "add_basket";
        } else {
          obj.user_defined.action_type = "filter_product";
        }
        acc.push(obj.user_defined);
      }
      return acc;
    }, []);
    data.action = actions[0];
  } catch (err) {
    console.log(err);
  }
  console.log(data);
  return data;
};

export const stripSSMLTags = (str) => {
  const regex = /(<([^>]+)>)/gi;
  try {
    str = str.replace(regex, "");
  } catch (err) {
    console.log(err);
  }
  return str;
};
