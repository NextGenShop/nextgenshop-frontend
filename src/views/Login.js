import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';
import Button from '@material-ui/core/Button';
import { Divider, TextField } from '@material-ui/core';
import { connect } from 'react-redux';
import { actions, types } from '../store/api/auth';
import { Redirect } from 'react-router-dom';
import { createErrorSelector } from '../store/error';
import { displayToast } from '../utils/displayToast';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: theme.spacing(2),
    paddingBottom: theme.spacing(2),
    height: '80vh',
  },
  heading: {
    paddingBottom: theme.spacing(2),
  },
  mr: {
    marginRight: theme.spacing(1),
  },
  mt: {
    marginTop: theme.spacing(1),
  },
}));

export function Login({ auth, error, dispatchLogin, dispatchClearErrors }) {
  const classes = useStyles();

  const [email, setEmail] = React.useState('');
  const [password, setPassword] = React.useState('');

  React.useEffect(() => {
    if (error) {
      displayToast('', error, 'danger', dispatchClearErrors);
    }
  }, [error, dispatchClearErrors]);

  const loginShopper = () => {
    dispatchLogin({ email: 'mockuser1@gmail.com', password: 'password' });
  };

  const loginCashier = () => {
    dispatchLogin({ email: 'mockcashieruser@gmail.com', password: 'password' });
  };

  const handleLogin = () => {
    if (email.trim() && password) {
      const data = { email, password };
      dispatchLogin(data);
    }
    setPassword('');
  };

  return auth && auth.isAuthenticated ? (
    <Redirect to='/' />
  ) : (
    <Container className={classes.root} maxWidth='md'>
      <Typography variant='h4' className={classes.heading}>
        NextGenShop Login
      </Typography>
      <TextField
        id='emailTextField'
        label='Email'
        value={email}
        onChange={(e) => setEmail(e.target.value)}
      />
      <TextField
        id='passwordTextField'
        type='password'
        label='Password'
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <Button
        id='loginButton'
        className={classes.mt}
        color='primary'
        variant='contained'
        onClick={handleLogin}
      >
        Login
      </Button>
      <Divider className={classes.mt} />
      <div className={classes.mt}>
        <Button
          id='loginShopperButton'
          className={classes.mr}
          variant='contained'
          onClick={loginShopper}
        >
          Login as shopper
        </Button>
        <Button
          id='loginCashierButton'
          variant='contained'
          onClick={loginCashier}
        >
          Login as cashier
        </Button>
      </div>
      <Typography className={classes.mt} variant='caption'>
        These buttons are for testing purposes.
      </Typography>
    </Container>
  );
}

const errorSelector = createErrorSelector([types.LOGIN]);

const mapStateToProps = (state) => ({
  auth: state.auth,
  error: errorSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  dispatchLogin: (data) => dispatch(actions.login(data)),
  dispatchClearErrors: () => dispatch(actions.clearErrors()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
