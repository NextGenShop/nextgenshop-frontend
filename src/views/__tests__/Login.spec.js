import React from 'react';
import { shallow, mount } from 'enzyme';
import { Login } from '../Login';
import { displayToast } from '../../utils/displayToast';

jest.mock('../../utils/displayToast', () => {
  return {
    displayToast: jest.fn(),
  };
});

describe('Login', () => {
  let props;
  beforeEach(() => {
    props = {
      auth: { isAuthenticated: false },
      error: null,
      dispatchLogin: jest.fn(),
      dispatchClearErrors: jest.fn(),
    };
  });
  test('should correctly render', () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });
  test('should correctly render with props', () => {
    const component = shallow(<Login {...props} />);
    expect(component).toMatchSnapshot();
  });
  test('should correctly login with email and password', () => {
    const email = 'mockEmail';
    const password = 'mockPassword';
    const component = shallow(<Login {...props} />);
    component
      .find('#emailTextField')
      .first()
      .simulate('change', {
        target: { value: email },
      });
    expect(component.find('#emailTextField').first().prop('value')).toEqual(
      email
    );
    component
      .find('#passwordTextField')
      .first()
      .simulate('change', { target: { value: password } });
    expect(component.find('#passwordTextField').first().prop('value')).toEqual(
      password
    );
    component.find('#loginButton').first().simulate('click');
    expect(props.dispatchLogin).toBeCalledWith({ email, password });
  });
  test('should correctly login as shopper', () => {
    const component = mount(<Login {...props} />);
    component.find('#loginShopperButton').first().simulate('click');
    expect(props.dispatchLogin).toBeCalledWith({
      email: 'mockuser1@gmail.com',
      password: 'password',
    });
  });
  test('should correctly login as cashier', () => {
    const component = mount(<Login {...props} />);
    component.find('#loginCashierButton').first().simulate('click');
    expect(props.dispatchLogin).toBeCalledWith({
      email: 'mockcashieruser@gmail.com',
      password: 'password',
    });
  });
  test('should display toast message on error', () => {
    props.error = 'mockError';
    const component = mount(<Login {...props} />);
    component.find('#loginShopperButton').first().simulate('click');
    expect(displayToast).toBeCalledWith(
      '',
      'mockError',
      'danger',
      props.dispatchClearErrors
    );
  });
});
