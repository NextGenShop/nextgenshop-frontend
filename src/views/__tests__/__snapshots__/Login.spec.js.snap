// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should correctly render 1`] = `
<WithStyles(ForwardRef(Container))
  className="makeStyles-root-1"
  maxWidth="md"
>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-heading-2"
    variant="h4"
  >
    NextGenShop Login
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(TextField))
    id="emailTextField"
    label="Email"
    onChange={[Function]}
    value=""
  />
  <WithStyles(ForwardRef(TextField))
    id="passwordTextField"
    label="Password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <WithStyles(ForwardRef(Button))
    className="makeStyles-mt-4"
    color="primary"
    id="loginButton"
    onClick={[Function]}
    variant="contained"
  >
    Login
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Divider))
    className="makeStyles-mt-4"
  />
  <div
    className="makeStyles-mt-4"
  >
    <WithStyles(ForwardRef(Button))
      className="makeStyles-mr-3"
      id="loginShopperButton"
      onClick={[Function]}
      variant="contained"
    >
      Login as shopper
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      id="loginCashierButton"
      onClick={[Function]}
      variant="contained"
    >
      Login as cashier
    </WithStyles(ForwardRef(Button))>
  </div>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-mt-4"
    variant="caption"
  >
    These buttons are for testing purposes.
  </WithStyles(ForwardRef(Typography))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Login should correctly render with props 1`] = `
<WithStyles(ForwardRef(Container))
  className="makeStyles-root-1"
  maxWidth="md"
>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-heading-2"
    variant="h4"
  >
    NextGenShop Login
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(TextField))
    id="emailTextField"
    label="Email"
    onChange={[Function]}
    value=""
  />
  <WithStyles(ForwardRef(TextField))
    id="passwordTextField"
    label="Password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <WithStyles(ForwardRef(Button))
    className="makeStyles-mt-4"
    color="primary"
    id="loginButton"
    onClick={[Function]}
    variant="contained"
  >
    Login
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Divider))
    className="makeStyles-mt-4"
  />
  <div
    className="makeStyles-mt-4"
  >
    <WithStyles(ForwardRef(Button))
      className="makeStyles-mr-3"
      id="loginShopperButton"
      onClick={[Function]}
      variant="contained"
    >
      Login as shopper
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      id="loginCashierButton"
      onClick={[Function]}
      variant="contained"
    >
      Login as cashier
    </WithStyles(ForwardRef(Button))>
  </div>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-mt-4"
    variant="caption"
  >
    These buttons are for testing purposes.
  </WithStyles(ForwardRef(Typography))>
</WithStyles(ForwardRef(Container))>
`;
